### PR TITLE
Fashion MNIST addition

### PR DESCRIPTION
**Pull request by subversive-owl**
- [x] `load-dataset fashion`
- [x] `fashion` command

### CODE REVIEW

**Review by drahnr — APPROVED** [review]

File: src/main.rs [hunk 1]
```diff
@@ -64,15 +64,15 @@ struct MainArgs {
     arg_momentum: Option<f32>,
     cmd_load_dataset: bool,
     cmd_mnist: bool,
+    cmd_fashion: bool
 }
```
Comment (drahnr) [inline src/main.rs]: Not sure what is the advantage of moving away from `Vec<u8>`

**Review by drahnr — APPROVED** [review]

**Review by drahnr — APPROVED** [review]
Nice work!

File: src/main.rs (hunk 1 above)
Comment (subversive-owl) [reply src/main.rs]: sorry where do you mean?

if it's `datasets: &[&str]`, that's just because that was their format in the code before refactoring.

**Comment by subversive-owl** [comment]
@drahnr Right now it runs, but it seems to be stuck reading one example over and over again. I'll look at it tomorrow, but maybe you'll spot the error.

File: src/main.rs (hunk 1 above)
Comment (drahnr) [reply src/main.rs]: Alright, I missed that and assumed it was `vec<u8>`, mea culpa. Leave as is :)

File: src/main.rs [hunk 2]
```diff
@@ -86,21 +88,25 @@ fn download_datasets(datasets: &[&str], base_url: &str) {
...
         let mut in_file: Vec<u8> = Vec::new();
+        let mut decompressed_file: Vec<u8> = Vec::new();
         
-        file_handle.read_to_end(&mut in_file);
+        file_handle.read_to_end(&mut in_file).unwrap();
+
+        let mut decoder = GzDecoder::new(in_file.as_slice()).unwrap();
 
-        let unzipped_data = inflate_bytes_zlib(in_file.as_slice()).unwrap();
+        decoder.read_to_end(&mut decompressed_file).unwrap();
```
Comment (drahnr) [inline src/main.rs]: There are a lots of `unwrap()`s, I think an `expect(...)` with some descriptive error text would be nicer.

**Review by drahnr — APPROVED** [review]

File: src/main.rs [hunk 3]
```diff
@@ -82,16 +84,16 @@ fn download_datasets(datasets: &[&str], base_url: &str) {
...
             .unwrap()
             .write_all(&body as &[u8]).unwrap();
     }
 }
 
-#[allow(unused_must_use)]
 fn unzip_datasets(datasets: &[&str]) {
     for filename in datasets {
-        let mut file_handle = File::open(&format!("assets/{}", filename))
+        // TODO figure out how to specify asset dir
```
Comment (drahnr) [inline src/main.rs]: use something like `std::env::var("JUICE_EXAMPLE_ASSET_DIR").unwrap_or("./assets".to_string())` unless you want to make it part of the commandline arguments.

**Comment by subversive-owl** [comment]
Docs around the IDX parser, for the record: https://docs.rs/mnist/0.4.0/mnist/

**Comment by subversive-owl** [comment]
still not fixed, but i learned some things about one of the iterators in the last commit...

here is the image decoder iterator for mnist CSV:

```rs
let mut decoded_images = rdr.decode().map(|row|
        match row {
            Ok(value) => {
                let row_vec: Box<Vec<u8>> = Box::new(value);
                let label = row_vec[0];
                let mut pixels = vec![0u8; pixel_count];
                for (place, element) in pixels.iter_mut().zip(row_vec.iter().skip(1)) {
                    *place = *element;
                }
                (label, pixels)
            },
            _ => {
                println!("no value");
                panic!();
            }
        }
    );
```

and then here is what i wrote for the fashion-mnist IDX output:

```rs
let mut decoded_images = trn_img.chunks(pixel_count).enumerate().map(|(ind, pixels)| {
        (trn_lbl[ind], pixels.to_vec())
    });
```

my thought was that you shouldn't need to zip over the image pixels themselves since they're already stored as `Vec<u8>` and are monochrome. if this code isn't right it's the iteration during the batch training loop below, but that's unchanged from before.

**Comment by subversive-owl** [comment]
Okay, it works!

**Comment by subversive-owl** [comment]
Is @sirmergealot going to merge this or is there something else that's supposed to happen?

**Pull request merged**